satpy/writers/__init__.py

```diff
@@ -232,56 +236,46 @@ def add_overlay(orig, area, coast_dir, color=(0, 0, 0), width=0.5, resolution=No
+                      " arguments will be deprecated soon. Please use 'overlays' instead.", DeprecationWarning)
 
-    if hasattr(orig, 'convert'):
+    if hasattr(orig_img, 'convert'):
         # image must be in RGB space to work with pycoast/pydecorate
-        orig = orig.convert('RGBA' if orig.mode.endswith('A') else 'RGB')
+        res_mode = ('RGBA' if orig_img.mode.endswith('A') else 'RGB')
+        orig = orig_img.convert(res_mode)
     elif not orig.mode.startswith('RGB'):
         raise RuntimeError("'trollimage' 1.6+ required to support adding "
                            "overlays/decorations to non-RGB data.")
-    img = orig.pil_image(fill_value=fill_value)
-    cw_ = ContourWriterAGG(coast_dir)
-    cw_.add_coastlines(img, area, outline=color,
-                       resolution=resolution, width=width, level=level_coast)
-    cw_.add_borders(img, area, outline=color,
-                    resolution=resolution, width=width, level=level_borders)
-    # Only add grid if major_lonlat is given.
-    if grid and 'major_lonlat' in grid and grid['major_lonlat']:
-        major_lonlat = grid.pop('major_lonlat')
-        minor_lonlat = grid.pop('minor_lonlat', major_lonlat)
 
-        cw_.add_grid(img, area, major_lonlat, minor_lonlat, **grid)
+    if overlays is None:
+        # fill with sensible defaults
+        general_params = {'outline': color or (0, 0, 0),
+                          'width': width or 0.5}
+        for key, val in general_params.items():
+            if val is not None:
+                overlays.setdefault('coasts', {}).setdefault(key, val)
+                overlays.setdefault('borders', {}).setdefault(key, val)
+        if level_coast is None:
+            level_coast = 1
+        overlays.setdefault('coasts', {}).setdefault('level', level_coast)
+        if level_borders is None:
+            level_borders = 1
+        overlays.setdefault('borders', {}).setdefault('level', level_borders)
+
+        if grid is not None and 'major_lonlat' in grid and grid['major_lonlat']:
+            major_lonlat = grid.pop('major_lonlat')
+            minor_lonlat = grid.pop('minor_lonlat', major_lonlat)
+            grid_params = {'Dlonlat': major_lonlat, 'dlonlat': minor_lonlat}
+            for key, val in grid_params.items():
+                overlays.setdefault('grid', {}).setdefault(key, val)
 
-    arr = da.from_array(np.array(img) / 255.0, chunks=CHUNK_SIZE)
-
-    new_data = xr.DataArray(arr, dims=['y', 'x', 'bands'],
-                            coords={'y': orig.data.coords['y'],
-                                    'x': orig.data.coords['x'],
-                                    'bands': list(img.mode)},
-                            attrs=orig.data.attrs)
-    return XRImage(new_data)
+    cw_ = ContourWriterAGG(coast_dir)
+    new_image = orig_img.apply_pil(_burn_overlay, res_mode,
+                                   None, {'fill_value': fill_value},
+                                   (area, cw_, overlays), None)
```

Comment:
Is it possible to provide backwards compatibility for a while or is it too much work?

Reply:
Should be doable, let's see

Reply:
ok, I can do this (copy apply_pil in satpy), but it's ugly and doesn't feel good as it's just a temporary patch.
Moreover, we also need a new version of pycoast for this to work.
So if you really want to keep backwards compatibility, I could just keep the old `add_overlay` and add a call to the new one inside in instead. But do we really need this ?

Reply:
> But do we really need this?

"Probably not" -Famous last words

Users are going to complain, but it seems like it is much easier on us to just say "update pycoast".